### PR TITLE
Yearn strategy fork tests

### PR DESCRIPTION
**Pull request by brkomir**
Testing yearn strategy contract on the mainnet fork to assert investing to the yearn vault works as expected, and as well in cases when yearn vault generates/looses money.

### CODE REVIEW

**Comment by brkomir** [comment]
> @brkomir can you change the base to `yearn_lusd_strat`? it's a bit hard to review like this

@gabrielpoca done

File: contracts/strategy/yearn/YearnStrategy.sol [hunk 1]
```diff
@@ -89,15 +89,20 @@ contract YearnStrategy is IStrategy, AccessControl, CustomErrors {
         return _getShares() > 0;
     }
 
-    /// @inheritdoc IStrategy
+    /**
```
Comment (gabrielpoca) [inline contracts/strategy/yearn/YearnStrategy.sol:92]: Why this change? This should be in IStrategy

Comment (brkomir) [reply contracts/strategy/yearn/YearnStrategy.sol]: My bad, [fixed](https://github.com/lindy-labs/sc_solidity-contracts/pull/139/commits/35122021e003fa1dc07af950485cf15406b0a436)

File: contracts/strategy/yearn/YearnStrategy.sol [hunk 2]
```diff
@@ -135,7 +136,9 @@ contract YearnStrategy is IStrategy, AccessControl, CustomErrors {
         }
 
         // transfer underlying to vault
-        underlying.safeTransfer(vault, amount);
+
+        underlying.safeIncreaseAllowance(address(this), amount);
```
Comment (gabrielpoca) [inline contracts/strategy/yearn/YearnStrategy.sol:140]: can't we just do a underlying.safeIncreaseAllowance MAX in the constructor and go with it? this is just transferring money between contracts cc @naps62

Comment (brkomir) [reply contracts/strategy/yearn/YearnStrategy.sol:140]: I suppose we can do this. I will go forward and make the change

Comment (brkomir) [reply contracts/strategy/yearn/YearnStrategy.sol]: [changed](https://github.com/lindy-labs/sc_solidity-contracts/pull/139/commits/7c1b86a53a9692aad901d0ba9640ddbd6d0d0a49)

File: contracts/strategy/yearn/YearnStrategy.sol [hunk 3]
```diff
@@ -57,6 +59,8 @@ contract YearnStrategy is IStrategy, AccessControl, CustomErrors {
         vault = _vault;
         yVault = IYearnVault(_yVault);
         underlying = IERC20(_underlying);
+
+        underlying.safeIncreaseAllowance(address(this), MAX_UINT256);
```
Comment (gabrielpoca) [inline contracts/strategy/yearn/YearnStrategy.sol:63]: let's change it to "approve" instead

File: contracts/interfaces/yearn/IYearnVault.sol [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.10;
```
Comment (naps62) [inline contracts/interfaces/yearn/IYearnVault.sol:2]: Why not use `^0.8.10` and make it easier to upgrade to 0.8.14? :)

or if you prefer strict versions, why not upgrade to 0.8.14 anyway?

File: contracts/strategy/yearn/YearnStrategy.sol (hunk 3 above)
Comment (naps62) [reply contracts/strategy/yearn/YearnStrategy.sol:63]: for the record, both should do the same, since this is a constructor (unless you use CREATE2, but let's not open that rabbit whole). but approve just feels more correct in this context

File: contracts/strategy/yearn/YearnStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        return
+            _sharesToUnderlying(_getShares()) + _getUnderlyingBalance();
+    }
+
+    /// @inheritdoc IStrategy
+    function invest() external virtual override(IStrategy) onlyManager {
+        uint256 balance = _getUnderlyingBalance();
+        if (balance == 0) revert StrategyNoUnderlying();
+
+        underlying.safeIncreaseAllowance(address(yVault), balance);
```
Comment (naps62) [inline contracts/strategy/yearn/YearnStrategy.sol:113]: this could also be moved to the constructor, right?

File: contracts/strategy/yearn/YearnStrategy.sol (hunk 3 above)
Comment (brkomir) [reply contracts/strategy/yearn/YearnStrategy.sol]: this fix is already on the main branch, [commit](https://github.com/lindy-labs/sc_solidity-contracts/commit/7aa480c0e96fe632f8c4396df420fec54aa0ca0e)

File: contracts/interfaces/yearn/IYearnVault.sol (hunk 4 above)
Comment (brkomir) [reply contracts/interfaces/yearn/IYearnVault.sol]: don't really have an answer for this, the same statement is in all contract files in the project

File: contracts/strategy/yearn/YearnStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    /// role allowed to invest/withdraw from yearn vault
+    bytes32 public constant MANAGER_ROLE = keccak256("MANAGER_ROLE");
+    // underlying ERC20 token
+    IERC20 public immutable underlying;
+    /// @inheritdoc IStrategy
+    address public immutable override(IStrategy) vault;
+    // yearn vault that this strategy is interacting with
+    IYearnVault public immutable yVault;
+    // uint256 max value
+    uint256 constant MAX_UINT256 = 2**256 - 1;
```
[the remark this answers is not in the document]
Comment (brkomir) [reply contracts/strategy/yearn/YearnStrategy.sol]: already fixed on main, [commit](https://github.com/lindy-labs/sc_solidity-contracts/commit/7aa480c0e96fe632f8c4396df420fec54aa0ca0e)
